### ISSUE TITLE
Try with failure::Error does not play nice with mutex (sync bound)

### ISSUE THREAD

**Issue by spease** [rust-lang-nursery/failure#192]
```
extern crate failure;

use failure::Error;
use std::sync::Mutex;

fn example() -> Result<(), Error> {
    let a = Mutex::new(());
    a.lock()?;
    Ok(())
}

fn main() {}
```

```
error[E0277]: the trait bound `std::sync::MutexGuard<'_, ()>: std::marker::Send` is not satisfied in `std::sync::PoisonError<std::sync::MutexGuard<'_, ()>>`
 --> src/main.rs:8:5
  |
8 |     a.lock()?;
  |     ^^^^^^^^^ `std::sync::MutexGuard<'_, ()>` cannot be sent between threads safely
  |
  = help: within `std::sync::PoisonError<std::sync::MutexGuard<'_, ()>>`, the trait `std::marker::Send` is not implemented for `std::sync::MutexGuard<'_, ()>`
  = note: required because it appears within the type `std::sync::PoisonError<std::sync::MutexGuard<'_, ()>>`
  = note: required because of the requirements on the impl of `failure::Fail` for `std::sync::PoisonError<std::sync::MutexGuard<'_, ()>>`
  = note: required because of the requirements on the impl of `std::convert::From<std::sync::PoisonError<std::sync::MutexGuard<'_, ()>>>` for `failure::Error`
  = note: required by `std::convert::From::from`

error: aborting due to previous error

error: Could not compile `failure_mutex`.

To learn more, run the command again with --verbose.
```

Any recommended way to deal with this? I suppose I could try wrapping the mutex error in a mutex... 🙃

Possibly more of an upstream issue. Should std::sync::PoisonError be made sync?

**Comment by dathinab** [comment]
TL;DR: I made the important parts bold

The problem is actually more then just send, i.e. `PoisonError` contains the `MutexGuard` which is not only not `Send` but also not `'static`!

There is a certain class(es) of errors which can not be covered by the current `Fail` design, this mainly includes errors, which contain a "value" as a form of context and/or additional metadata (e.g. from a failed conversion which consumed `self`, a `&str` slice pointing to what part of parsing failed etc.).

The problem is that fixing it would require to remove the `Send`, `Sync` and `'static` bound from `Fail` and while this can be done[*] it would negate many of the usability benefits of `Fail` and therefore won't be done.

The solution for this is to accept that **the `failure` crate is not suited for all kinds of error cases**,
luckily, **the cases it is not good at are such which tend to be not bubbled upwards anyway**.

Even if it would be possible I would strongly discourage using any form of crate/library level
error propagation with a `PoisonError` in most (but not all) cases, a lock being poisoned
often means there is not a error but a Bug (if seen from a crate level perspective, from the
perspective of the lock api it is indeed a error) so I would do one of following thinks:

1. If you know poisoning can not lead to corruption due to the way the lock is used
    you might ignore poison, i.e. you don't need any error propagation (_only do so
    if you are sure you understand what you are doing [**]_)

2. **Create your own lock poison error** possible with more context (e.g.
    in it's display text or name) about what was poisoned and how this could
   have happened.
3. If you know only your code could cause it but should not, `panic` can be
    fine too, you basically reached a branch of execution which should not be 
    be reachable, still if it's a library panic might also be suboptimal

---

([*] one problem with `std::error::Error` is the cause not being static and therefore not being  castable, BUT this can be fixed by stating in the signature of `cause` that the returned  value  has to live at last as long as this error, i.e. if self: 'static then the returned Fail is static too)

([**] There are a bunch of ways in which panic+recovery can wrack havoc, even worse while
 code using unsafe has to consider panic, i.e. it is not allowed to become unsafe just because it was
 involved in a panic, it's quite likely that there is a lot of code which does not do so 100% correctly.)

---

Here a simple custom poison error:

```rust
#[derive(Debug, Fail)]
#[fail(display = "resource bla Mutex was poisoned")]
pub struct ResourceBlaPoisonError {
    backtrace: Backtrace
}

impl ResourceBlaPoisonError {
    pub fn new() -> ResourceBlaPoisonError {
        ResourceBlaPoisonError {
            backtrace: Backtrace::new()
        }
    }
}

//for op-?, "auto" type conversion
impl<T> From<sync::PoisonError<T>> for ResourceBlaPoisonError {
    fn from(_: sync::PoisonError<T>) -> Self {
        ResourceBlaPoisonError::new()
    }
}
```

**Comment by dathinab** [comment]
Also:

_Don't fall into the fallacy to think that just because there is a `Fail`/`Error` trait every
  error has to implement it_

(through if you do so 1. have a good reason for it and 2. provide a way to convert it to
a `Fail`/`Error` Error)
